chassis/core/src/main/java/com/griddynamics/jagger/invoker/ExclusiveAccessRandomLoadBalancer.java

```diff
@@ -0,0 +1,104 @@
+package com.griddynamics.jagger.invoker;
+
+import com.griddynamics.jagger.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.AbstractIterator;
+
+import java.util.Iterator;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This {@link com.griddynamics.jagger.invoker.LoadBalancer} implementation has the same semantics
+ * as {@link com.griddynamics.jagger.invoker.RandomLoadBalancer} and in addition it guarantees
+ * that each query and endpoint pair will be in exclusive access, i.e. once it is acquired by one thread
+ * it won't be acquired by any other thread (virtual user).
+ * @n
+ * Created by Andrey Badaev
+ * Date: 01/02/17
+ */
+public class ExclusiveAccessRandomLoadBalancer<Q, E> extends PairSupplierFactoryLoadBalancer<Q, E> {
+    
+    private final static Logger log = LoggerFactory.getLogger(ExclusiveAccessRandomLoadBalancer.class);
+    
+    public ExclusiveAccessRandomLoadBalancer(long randomSeed, PairSupplierFactory<Q, E> pairSupplierFactory) {
+        super(pairSupplierFactory);
+        this.randomSeed = new AtomicLong(randomSeed);
+    }
+    
+    private final AtomicLong randomSeed;
+    
+    public long getRandomSeed() {
+        return randomSeed.get();
+    }
+
+    private volatile ConcurrentHashMap<Integer, Pair<Q, E>> pairMap;
+    private volatile int pairsNumber;
+    
+    @Override
+    public Iterator<Pair<Q, E>> provide() {
+        return new AbstractIterator<Pair<Q, E>>() {
+    
+            private Random random = new Random(randomSeed.getAndIncrement());
+            
+            private Pair<Q, E> currentPair;
+            private Integer currentIndex;
+    
+            @Override
+            protected Pair<Q, E> computeNext() {
+    
+                if (currentPair != null && currentIndex != null) {
+                    log.debug("Returning pair - {}", currentPair);
+                    pairMap.put(currentIndex, currentPair);
+                }
+                
+                long startTime = System.currentTimeMillis();
+                long maxDurationInMinutes = 10;
+                do {
+                    currentIndex = random.nextInt(pairsNumber);
+                    currentPair = pairMap.remove(currentIndex);
```

Comment:
You should reduce the number of pairs each time when you take a pair from the queue.